### PR TITLE
multi-tenancy CRD design

### PR DESCRIPTION
**Pull request by chenqianfzh**
doc update.

### CODE REVIEW

**Comment by chenqianfzh** [comment]
> That's a cool document.
> 
> I have a question: consider this example: "An Operator creates a system-tenant CRDs. Initially, the objects of these CRDs are not visible to other tenants, would it be possible to transfer objects of a system's CRD to a tenant's scope and the objects remain manageable by the operator?"

Qian: Yes, it is doable. A cluster operator is a super admin. However, it can never be a recommended behavior.

File: docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md [hunk 1]
```diff
@@ -159,14 +159,88 @@ Here are some examples:
...
 
- Each tenant can install their CRD independently, without worrying about resource name collisions. The API group/vision/kind (G/V/K) defined by the CRD object is only accessible to that tenant.
+#### Principles and Outlines of Design
+
+ The following Arktos multi-tenancy design principles will be observed in the design of CRD:
  
- For CRDs installed in system world, by default it's only accessible for tenant users. But if it's marked with "multi-tenancy.k8s.io/published" annotation, it will be accessible to all tenants.
+ 1. **isolation**. Each regular tenant can independently create/delete/update their CRDs and custom resources based on the CRDs. There will be NO CRD G/V/K collision between two regular tenants. Actually, the CRDs and custom resources of one regular tenant are invisible to another regular tenant.  
+
+ 2. **Managebility**. A cluster operator (he should be user under the system tenant) can introduce a new CRD, where he can choose to:
```
Comment (XiaoningDing) [inline docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md:168]: If the design of manageability part is not clear we can skip it for now.

File: docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md [hunk 2]
```diff
@@ -159,14 +159,88 @@ Here are some examples:
...
+  * make it usable only to the system tenant.
+  * or, make it usable only to a spefic set of to regular tenants and the system tenants, 
+  * or, usable to all the regular tenants. 
+  * Additionally, the applicable range of a CRD can be changed by the cluster operator at any time in a simple and quick manner. That will be useful to CRD test and rolling update.  
  
- If there is a collision between CRDs in system world and tenant world, the latter one takes precedence if it's accessed by a tenant.
+ Note: it does not contradict the principle of isolation. The isolation principle is about two regular tenants, while this principle is about the system tenant and regular tenants. 
+
+ 3. **Autonomy**. The tenant admin has the final say on what CRD will be used in his tenant space when there is an overlap between the tenant's CRD and that of the system tenant. The tenant can choose MultiTenancyCRD policy to be :
```
Comment (XiaoningDing) [inline docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md:176]: Autonomy is also about a tenant can install his own CRDs without help from cluster admins.

File: docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md [hunk 3]
```diff
@@ -159,14 +159,88 @@ Here are some examples:
...
+
+ 4. **Backward compatibility**. The legacy CRD definitions and CRD operators continue to work in Arktos. More specifically, it means:
+    * An existing defintion of CRD (usually in a .yaml file) can be applied without any change.
+    * An existing CRD operator image can work in Arktos without re-compiling or image rebuilding.
+    * An existing CRD operator source code build and work in Arktos. So developers can create new operators or revise existing operators in Arktos without learning new APIs.
+
+However, we need to point out that, Arktos does not allow a regular tenant to access cluster-scope resources, such as nodes and daemonsets. So a CRD definition or a CRD operator that needs access to the cluster-scope resources will no longer be supported in a regular tenant's space. Surely, they can continue to work in the system tenant's space. For the same reason, a regular tenant is not allowed to create a cluster-scope CRD. If regular tenant specifies its CRD to be cluster-scoped, the CRD will be changed to tenant-scoped.
```
Comment (XiaoningDing) [inline docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md:190]: Suggest rewording to something like "certain resources are not accessible from regular tenants, like node objects".   "Arktos does not allow a regular tenant to access cluster-scope resources..." is too strong since there are many cluster-scoped resources.

File: docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md [hunk 4]
```diff
@@ -159,14 +159,88 @@ Here are some examples:
...
+ 
+#### Detailed Design
+
+ The key changes to realize the isolation principles are:
+ * CustomResourceDefinition will be changed to a tenant-scoped resource. As a result, same CRD G/V/K can co-exist as long as they live under different tenants. 
+ * APIResource struct ( which is the data structure to holds the info of resource types suspported in API server) will have a new field, Tenant. For non-CRD resources, this field is empty, which means the resources are applicable to all tenants. For CRD resources, it indicates the belonging tenant.
+ * The CRD shown in the response to a resource discovery request should only show those usable to tenant in the request.
 
+The key changes to realize managability and automony are:
+* A new field, ApplicableRange, will be added to the CustomResourceDefinitionSpec. The field is ignored in a regular-tenant CRD. Yet for a system-tenant CRD, it indicates the scope of the regular tenants that this CRD is applicable to. This field is a list of annotations, which are used to finding matching tenants. (other designs may also work, as long as the field give info in finding matching tenants)
```
Comment (XiaoningDing) [inline docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md:201]: This only solves the problem of discovering CR resources. But it requires code changes in operator to handle multi-tenancy CRs.

File: docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md [hunk 5]
```diff
@@ -159,14 +159,88 @@ Here are some examples:
 
 ### CRD Support
```
Comment (XiaoningDing) [inline docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md:160]: Some general comments:  1) consider leaving out uncertain parts and only check in the finalized part.  2) there are some minor grammar mistakes or typos, suggest running spelling and grammar check in editor before final commit.

File: docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md (hunk 1 above)
Comment (chenqianfzh) [reply docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md:168]: Add some notes that this part is still under update.

File: docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md (hunk 2 above)
Comment (chenqianfzh) [reply docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md:176]: Yep, actually this point is more point. Added it as the first line in the subsection of "autonomy".

File: docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md (hunk 3 above)
Comment (chenqianfzh) [reply docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md:190]: updated as suggested.

File: docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md (hunk 4 above)
Comment (chenqianfzh) [reply docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md:201]: Yes, the existing operator only works in a back-compatible way within his own tenant'world. If an operator needs access multi-tenant CRs, changes are inevitable. 

I added a few lines in "backward-compatibility" to highlight this.

File: docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md (hunk 5 above)
Comment (chenqianfzh) [reply docs/design-proposals/multi-tenancy/multi-tenancy-api-resource-model.md:160]: Done.
